From f0a204ec13992d8c39f8624864222d49593e10a2 Mon Sep 17 00:00:00 2001
From: theofficialgman <28281419+theofficialgman@users.noreply.github.com>
Date: Sat, 2 May 2026 22:15:15 -0400
Subject: [PATCH 5/6] keep dimensions display native even if userspace attempts
 to swap them

userspace may attempt to check or set swapped dimensions due to having the dimensions switched in tegra_fb_modedb.

without this, not working in the display's native dimensions for fb can cause an out of bounds read/write leading to a kernel panic
Signed-off-by: theofficialgman <dofficialgman@gmail.com>
---
 drivers/video/tegra/fb.c | 38 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 38 insertions(+)

diff --git a/nvidia/drivers/video/tegra/fb.c b/nvidia/drivers/video/tegra/fb.c
index b46aa983c..48030fdd1 100644
--- a/nvidia/drivers/video/tegra/fb.c
+++ b/nvidia/drivers/video/tegra/fb.c
@@ -168,6 +168,25 @@ static int tegra_fb_check_var(struct fb_var_screeninfo *var,
 	struct tegra_dc *dc = tegra_fb->win.dc;
 	struct tegra_dc_out_ops *ops = dc->out_ops;
 	struct fb_videomode mode;
+	u32 tmp;
+
+	if (dc->out && (dc->out->rotation == 90 || dc->out->rotation == 270)) {
+		if (var->xres > var->yres) {
+			/* Swap dimensions back to native if they have been modified by userspace */
+			tmp = var->xres;
+			var->xres = var->yres;
+			var->yres = tmp;
+			tmp = var->left_margin;
+			var->left_margin  = var->upper_margin;
+			var->upper_margin = tmp;
+			tmp = var->right_margin;
+			var->right_margin  = var->lower_margin;
+			var->lower_margin  = tmp;
+			tmp = var->hsync_len;
+			var->hsync_len = var->vsync_len;
+			var->vsync_len = tmp;
+		}
+	}
 
 	if (tegra_fb->valid &&
 		(var->yres * var->xres * var->bits_per_pixel /
@@ -207,6 +226,25 @@ static int tegra_fb_set_par(struct fb_info *info)
 	struct tegra_fb_info *tegra_fb = info->par;
 	struct fb_var_screeninfo *var = &info->var;
 	struct tegra_dc *dc = tegra_fb->win.dc;
+	u32 tmp;
+
+	if (dc->out && (dc->out->rotation == 90 || dc->out->rotation == 270)) {
+		if (var->xres > var->yres) {
+			/* Swap dimensions back to native if they have been modified by userspace */
+			tmp = var->xres;
+			var->xres = var->yres;
+			var->yres = tmp;
+			tmp = var->left_margin;
+			var->left_margin  = var->upper_margin;
+			var->upper_margin = tmp;
+			tmp = var->right_margin;
+			var->right_margin  = var->lower_margin;
+			var->lower_margin  = tmp;
+			tmp = var->hsync_len;
+			var->hsync_len = var->vsync_len;
+			var->vsync_len = tmp;
+		}
+	}
 
 	if (var->bits_per_pixel) {
 		/* we only support RGB ordering for now */
-- 
2.43.0

